python-package/lightgbm/callback.py

```diff
@@ -63,15 +76,16 @@ def print_evaluation(period=1, show_stdv=True):
     callback : function
         The callback that prints the evaluation results every ``period`` iteration(s).
     """
-    def _callback(env):
+    @callback_attr_decorator
+    def _callback(env: CallbackEnv) -> None:
         if period > 0 and env.evaluation_result_list and (env.iteration + 1) % period == 0:
             result = '\t'.join([_format_eval_result(x, show_stdv) for x in env.evaluation_result_list])
             _log_info('[%d]\t%s' % (env.iteration + 1, result))
     _callback.order = 10
     return _callback
 
 
-def record_evaluation(eval_result):
+def record_evaluation(eval_result: dict) -> CallbackWithAttributes:
```

Comment:
```suggestion
def record_evaluation(eval_result: Dict[str, Dict[str, List[Any]]]) -> CallbackWithAttributes:
```

I think this could be a bit more specific, and being more specific will get us more informative errors from `mypy`.